File creation for apiserver arguments with file paths

srm09 opened the pull request:
**What this PR does / why we need it**:

1. This PR adds the functionality that allows the kubo release users to send a key value pair (flag name and encoded file content) which will be written out to files and the corresponding flags are set with the local path of the file that was created on the VM.
2. This also removes the unused `encryption-config` property spec from the kube-apiserver job.

larham commented on src/file-generator/generator/file.go:
```diff
@@ -0,0 +1,78 @@
+type FileGenerator struct {
+	FileName          string
+	ComponentName     string
+	FilePathGenerator PathGenerator
```
Is FilePathGenerator the output location?  Can its name be distinguished from the input?

larham commented on jobs/kube-apiserver/spec:
```diff
@@ -1,13 +1,15 @@
 name: kube-apiserver
 templates:
   bin/ensure_apiserver_healthy.erb: bin/ensure_apiserver_healthy
+  bin/generate-argument-files: bin/generate-argument-files
```
consider name "generate-config-files" since that is the output, although I see that you are reusing the nice input naming of "file-arguments".   Naming is hard :)

larham commented on jobs/kube-apiserver/spec:
```diff
@@ -40,6 +42,11 @@ properties:
+    description: "Pass-through options for Kubernetes runtime arguments which accept local file paths as inputs. See docs https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/ for reference."
+    example: |
+      file-arguments:
+        encryption-provider-config: blah-blah-blah-yaml
```
for an example, consider something that looks more like file content, perhaps

```
base64 of ----BEGIN CERTIFICATE------\nblah\n----END CERTIFICATE------
```

calebamiles commented on src/file-generator/generator/file.go:
```diff
@@ -0,0 +1,78 @@
+	}
+	defer jsonFile.Close()
+
+	byteValue, _ := ioutil.ReadAll(jsonFile)
```
I imagine you won't swallow this error in the final version of the PR

srm09 commented:
@larham @calebamiles Addressed the review comments and added tests for the code. Most of the file are vendor files and have very little to do with the actual change.

larham commented:
Nice update.  However, this commit now has 600 files involved because of updating vendored packages.

What do you think of separating out the vendor update?  I would be nice to see vendors get green separately, as a push to master directly, to rule out any issue caused by updating dependencies.

larham commented on src/file-generator/generator/file_test.go:
```diff
@@ -0,0 +1,104 @@
+
+			for _, flag := range []string{"encryption-provider-config", "oidc-ca-file"} {
+				generatedFilePath := filepath.Join(dirPath, jobName, flag)
+				Expect(generatedFilePath).To(BeAnExistingFile())
```
Consider asserting about the content of the file.

larham commented on src/file-generator/generator/file.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+func (f FileGenerator) writeToDisk(flag, content string) error {
+	decodedContent, err := base64.StdEncoding.DecodeString(content)
```
A test of file content should also cover this line since it can assert that that we decoded the base64 expected within the file

srm09 replied:
Done

daniel-keeney commented on packages/file-generator/packaging:
```diff
@@ -0,0 +1,15 @@
+
+# Create the executable file
+mkdir -p ${BOSH_INSTALL_TARGET}/bin
+go build -o ${BOSH_INSTALL_TARGET}/bin/file_generator ${BOSH_COMPILE_TARGET}/src/file-generator/main.go
```
Can this either be changed to add `test` to the commands (i.e. `go test build`), or add a separate line above it which will run the unit tests (and fail this script) before compiling?  Example here:
https://github.com/pivotal-cf/pks-kubo-release/blob/master/packages/osl-collector/packaging#L10-L11

If I overlooked where the tests are already included as part of CI, then please disregard this.

srm09 replied:
This is something that I needed some input on, as to where do I run these tests in the BoshCycle CI. I don't think the tests need to be run as part of the packaging script, since these sanity tests should be run even before a release is made.

daniel-keeney replied:
We had a chore to run unit tests as part of the `cfcr-dependencies` pipeline ( https://www.pivotaltracker.com/story/show/170775860 ), but that might not be very quick for you to pick up at the moment.  I agree with your overall point that tests should be run before you even make a release, but we are already following that practice for another package in this repo, so it might be the fastest way to get this out.

@larham do you have any preferences for how we can automatically run unit tests?

larham replied:
@srm09 yes, I agree that the packaging isn't the best place to run tests, but it may be fine because these tests are fast. It is a common practice to use `go test build`. Furthermore, there is a chance that disconnecting the test step from the build step could lead to a serious mistake that the tests were never run.

srm09 replied:
Added to the packaging script.

daniel-keeney commented:
We are already vendoring dependencies for the smoke-test package, so it seems that this follows the same convention.  An alternative could be to have the compiled binary as a bosh dependency, and have it be built as a separate repo, but that might be more work than we have time for with the alpha candidate coming up.

Overall, this seems fine after addressing running the unit tests as part of compilation

larham commented on jobs/kube-apiserver/templates/bin/generate-config-files:
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash -exu
```
Note that for kubelet, which runs on Windows, a matching .ps1 script might be necessary

srm09 replied:
Yes I had this conversation with @tenczar too, and something I will make a note of.

larham commented on src/file-generator/generator/file.go:
```diff
@@ -0,0 +1,70 @@
+func (f FileGenerator) writeToDisk(flag, content string) error {
+	decodedContent, err := base64.StdEncoding.DecodeString(content)
+	if err != nil {
+		return err
```
The code relies on the caller to submit only base64-encoded content. 
 Consider that a failure to decode base64 might look like

```
error: illegal base64 data at input byte 16
```
Is there any chance of the user not understanding this? If so, consider wrapping this error with a comment like "Expected input format, base64, not met: <error>"

srm09 replied:
This would not be a user facing error, the most that it would be is an error that could be found by alana after combing the logs. Still seems like a good idea to wrap the error into something more meaningful and user-friendly.

srm09 replied:
Done
